src/test/java/com/onlinebookstore/onlinebookstore/service/impl/BookServiceImplTest.java

```diff
@@ -0,0 +1,297 @@
+
+        assertNotNull(savedBookDto);
+        assertEquals(dto.getTitle(), savedBookDto.getTitle());
+    }
+
+    @Test
+    @DisplayName("Update existing Book")
+    @Sql(scripts = {
+            "classpath:database/book/remove-all-from-books-table.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    void changeExistingBook_ValidData_OK() {
+        Book book = new Book();
+        book.setId(1L);
+        book.setTitle("Book 1");
+
+        BookRequestDto requestDto = new BookRequestDto();
+        requestDto.setTitle("Book 2");
+        Set<Long> categoryIds = new HashSet<>();
+        categoryIds.add(1L);
+        requestDto.setCategoryIds(categoryIds);
+
+        BookResponseDto bookResponseDto = new BookResponseDto();
+        bookResponseDto.setId(1L);
+        bookResponseDto.setTitle("Book 2");
+
+        Mockito.when(bookRepository.findById(1L)).thenReturn(Optional.of(book));
+        List<Category> categories = categoryIds.stream()
+                .map(id -> {
+                    Category category = new Category();
+                    category.setId(id);
+                    return category;
+                })
+                .collect(Collectors.toList());
+        Mockito.when(categoryRepository.findAllById(categoryIds)).thenReturn(categories);
+        Mockito.when(bookRepository.save(book)).thenReturn(book);
+        Mockito.when(bookMapper.toDto(book)).thenReturn(bookResponseDto);
+
+        BookResponseDto updatedBookDto = bookService.update(1L, requestDto);
+
+        Set<Category> categorySet = new HashSet<>(categories);
+        assertEquals(categorySet, book.getCategories());
+
+        assertNotNull(updatedBookDto);
+        assertEquals(requestDto.getTitle(), updatedBookDto.getTitle());
+    }
+
+    @Test
+    @DisplayName("Find book by book id")
+    @Sql(scripts = {
+            "classpath:database/book/remove-all-from-books-table.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    void findBookById_ValidData_OK() {
+        Book book = new Book();
+        book.setId(1L);
+        book.setTitle("Book 1");
+
+        BookResponseDto dto = new BookResponseDto();
+        dto.setId(1L);
+        dto.setTitle("Book 1");
+
+        Mockito.when(bookRepository.findById(1L)).thenReturn(Optional.of(book));
+        Mockito.when(bookMapper.toDto(book)).thenReturn(dto);
+
+        BookResponseDto foundBook = bookService.findById(1L);
+
+        assertNotNull(foundBook);
+        assertEquals(dto.getId(), foundBook.getId());
+        assertEquals(dto.getTitle(), foundBook.getTitle());
+    }
+
+    @Test
+    @DisplayName("Delete book by id")
+    @Sql(scripts = {
+            "classpath:database/book/remove-all-from-books-table.sql"
+    }, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    void deleteBookById_ValidData_OK() {
+        Book book = new Book();
+        book.setId(1L);
+        book.setTitle("Book 1");

```

Comment:
Code duplication

Reply:
Actually there are a lot of duplication when you create mock objects it makes sense to make a methods for that